### PR TITLE
Modernize QgsQuick code

### PR DESCRIPTION
**Pull request by m-kuhn**
- RAII
- Enable render cache
- ~Removes QgsQuickUtils which is no longer needed~
- Improve stability by catching exceptions
- Add support for map canvas background color
- Add methods to convert from map units to device pixels (to avoid working directly with physical pixels)
- Saves battery by cancelling unneeded rendering jobs
- Reacts smoother to screen geometry changes (e.g. on device screen rotation)
- Works with transform context

This brings the QgsQuick map rendering engine up  to date with QField

### CODE REVIEW

File: src/quickgui/qgsquickmapcanvasmap.cpp [hunk 1]
```diff
@@ -14,26 +14,26 @@
...
 #include "qgsquickmapcanvasmap.h"
 #include "qgsquickmapsettings.h"
-#include "qgsexpressioncontextutils.h"
 
 
 QgsQuickMapCanvasMap::QgsQuickMapCanvasMap( QQuickItem *parent )
   : QQuickItem( parent )
-  , mMapSettings( new QgsQuickMapSettings() )
+  , mMapSettings( std::make_unique<QgsQuickMapSettings>() )
+  , mCache( std::make_unique<QgsMapRendererCache>() )
```
Comment (m-kuhn) [inline src/quickgui/qgsquickmapcanvasmap.cpp:36]: AFICS that was always null before and disfunctional?

File: src/quickgui/qgsquickmapcanvasmap.cpp [hunk 2]
```diff
@@ -320,15 +333,12 @@ QSGNode *QgsQuickMapCanvasMap::updatePaintNode( QSGNode *oldNode, QQuickItem::Up
 
 void QgsQuickMapCanvasMap::geometryChanged( const QRectF &newGeometry, const QRectF &oldGeometry )
 {
-  Q_UNUSED( oldGeometry )
-  // The Qt documentation advices to call the base method here.
```
Comment (m-kuhn) [inline src/quickgui/qgsquickmapcanvasmap.cpp:324]: Seems to be stable on QField side

File: src/quickgui/qgsquickmapcanvasmap.h [hunk 3]
```diff
@@ -16,25 +16,24 @@
...
 
 class QgsMapRendererParallelJob;
 class QgsMapRendererCache;
 class QgsLabelingResults;
 
 /**
  * \ingroup quick
- * \brief This class implements a visual Qt Quick Item that does map rendering
+ *
+ * This class implements a visual Qt Quick Item that does map rendering
```
Comment (m-kuhn) [inline src/quickgui/qgsquickmapcanvasmap.h:36]:
```suggestion
 * \brief This class implements a visual Qt Quick Item that does map rendering
```

**Comment by PeterPetrik** [comment]
Utils are used in the https://github.com/qgis/QGIS/blob/7f69678697ca65186682bfeb34beb3d72e07b526/src/quickgui/plugin/qgsquickmapcanvas.qml#L53 and I think in general you need `dp` in any qml component?

**Comment by m-kuhn** [comment]
With quickcontrols2 you no longer need this

**Comment by m-kuhn** [comment]
Even without it I think you would be better off taking the density of the screen the window is shown on and not a eandom singleton one.

File: src/quickgui/qgsquickmapcanvasmap.cpp (hunk 1 above)
Comment (nirvn) [reply src/quickgui/qgsquickmapcanvasmap.cpp:36]: @m-kuhn , correct, that was my finding when implementing cache a while back.

**Comment by PeterPetrik** [comment]
> With quickcontrols2 you no longer need this

interesting, what do you mean by it?

if it it is not needed, maybe we can drop utils...

**Comment by m-kuhn** [comment]
This bit here: https://doc.qt.io/qt-5/qtquickcontrols2-highdpi.html

> Warning: Qt Quick Controls 1.x are not compatible with Qt's automatic high-DPI scaling. There is currently no available solution for high-DPI support in applications wishing to mix Qt Quick Controls and Qt Quick Controls.

> if it it is not needed, maybe we can drop utils...

The same value is also available from `mapSettings.outputDpi` for the screen where the app runs on.
I've left quickutils and other methods like `mapUnitsPerPixel` in place, so the api does not require apps to run in high dpi mode, on QField side these are no longer needed.

File: src/quickgui/qgsquickmapsettings.h [hunk 4]
```diff
@@ -79,7 +82,7 @@ class QUICK_EXPORT QgsQuickMapSettings : public QObject
     /**
      * The background color used to render the map
      *
-     * The value is set to the project's background color setting on QgsProject::readProject
+     * The value is set to the project's bacckground color setting on QgsProject::readProject
```
Comment (PeterPetrik) [inline src/quickgui/qgsquickmapsettings.h:85]: typo

Comment (m-kuhn) [reply src/quickgui/qgsquickmapsettings.h:85]:
```suggestion
     * The value is set to the project's background color setting on QgsProject::readProject
```

**Comment by PeterPetrik** [comment]
Thanks for the update! we will look at the removal of the utils.dp in Input and potentially remove it from QGIS. At least for now the utils is only class where we have "prototype" of testing framework for this library...

**Review by PeterPetrik — APPROVED** [review]

**Comment by m-kuhn** [comment]
Exciting, thanks for the review @PeterPetrik

**Pull request merged**